client/extensions/woocommerce/app/dashboard/pre-setup-view.js

```diff
@@ -10,46 +10,100 @@ import React, { Component, PropTypes } from 'react';
  * Internal dependencies
  */
 import AddressView from 'woocommerce/components/address-view';
+import {
+	areSettingsGeneralLoading,
+	getStoreLocation,
+} from 'woocommerce/state/sites/settings/general/selectors';
+import { errorNotice } from 'state/notices/actions';
+import { getSelectedSiteWithFallback } from 'woocommerce/state/sites/selectors';
+import { fetchSettingsGeneral, updateStoreAddress } from 'woocommerce/state/sites/settings/general/actions';
+import { getSiteTitle } from 'state/sites/selectors';
+import { setSetStoreAddressDuringInitialSetup } from 'woocommerce/state/sites/setup-choices/actions';
 import SetupFooter from './setup-footer';
 import SetupHeader from './setup-header';
 
-import {
-	setSetStoreAddressDuringInitialSetup,
-} from 'woocommerce/state/sites/setup-choices/actions';
-
 class PreSetupView extends Component {
+	constructor( props ) {
+		super( props );
+		this.state = {
+			address: props.address,
+			isSaving: false,
+			userBeganEditing: false,
+		};
+	}
+
 	static propTypes = {
 		site: PropTypes.shape( {
 			ID: PropTypes.number.isRequired,
 		} ),
 	};
 
-	onChange = ( /* event */ ) => {
-		// TODO - store changes in state
+	componentDidMount = () => {
+		const { site } = this.props;
+
+		if ( site && site.ID ) {
+			this.props.fetchSettingsGeneral( site.ID );
+		}
+	}
+
+	componentWillReceiveProps = ( newProps ) => {
+		const { site } = this.props;
+		const newSiteId = site.selectedSite ? newProps.selectedSite.ID : null;
+		const oldSiteId = site ? site.ID : null;
+
+		if ( newSiteId && ( oldSiteId !== newSiteId ) ) {
+			this.props.fetchSettingsGeneral( newSiteId );
+		}
+
+		if ( ! this.state.userBeganEditing ) {
+			this.setState( { address: newProps.address } );
+		}
+	}
+
+	onChange = ( event ) => {
+		const address = this.state.address;
+		address[ event.target.name ] = event.target.value;
+		this.setState( { address, userBeganEditing: true } );
 	}
 
 	onNext = ( event ) => {
-		// TODO - save the address to the site
+		const { site, translate } = this.props;
 		event.preventDefault();
-		// TODO - maybe make button disabled while saving
-		this.props.setSetStoreAddressDuringInitialSetup( this.props.site.ID, true );
+		this.setState( { isSaving: true } );
+
+		// TODO before attempting to set the address, make sure all required
+		// plugins are installed and activated
+
+		const onSuccess = () => {
+			this.setState( { isSaving: false } );
+			return setSetStoreAddressDuringInitialSetup( this.props.site.ID, true );
+		};
+
+		const onFailure = () => {
+			this.setState( { isSaving: false } );
+			return errorNotice( translate( 'There was a problem saving the store address. Please try again.' ) );
+		};
+
+		this.props.updateStoreAddress(
+			site.ID,
+			this.state.address.street,
```

Comment:
I was a little nervous relying on spread operator here - to keep the ordering in the call should the object construction change